### PR TITLE
New "Boss Mod" triggers working with both DBM & BigWigs

### LINKED ISSUE #4502: Improving and Unifying bossmod callbacks
This isn't an overnight thing and something to be worked on by next tier or at least by next expansion. I have many notes on it here:
https://github.com/DeadlyBossMods/DBM-Unified/issues/231#issuecomment-1596023699

But basic gist is, BW and DBM callbacks are mostly cross compatible as is, and for majority of auras that aren't using a count or stupidly using msg text scanning when they could be using spellID, they work 1 for 1 just fine. However, when an author writes a weak aura, they target either bw or DBM and not both. It results in a weak aura that COUNT work with both mods, but doesn't.

There are a few steps to be able to unify the callbacks and they are basically
1. Get on same page with stage callback (bw and DBM problem not WA)
2. Support count arg in timer and announce objects so spellid + count can be used instead of string matching (DBM just did this, so did BW. It's better this is done in boss mods and not as reliant for weak aura's to do it in your handler cause this also just benefits all addons better or WAs using custom code instead of built in trigger parsing.)
3. BW needs to improve type filter. DBM lets you filter by ALL timer types, but bw is just "cast" and "cd" which means an author of a weak aura still needs to scan timer text itself for any timer object not using a spellID (such as a stage timer). This is the reason why the DBM type filter lets you choose a multitude of types. A good weak aura should be able to avoid requiring localized text scanning.

What needs doing on WA short term:
1. Adding support for the recently added count filter i just pushed to DBM. (https://github.com/DeadlyBossMods/DBM-Unified/commit/a9dae8eab96be1eb23b2001bd22710728ae5658a)
2. More stuff as BW adds it?
3. Scrap the "timer id" field on DBM timer start config. It's confusing and should never be used here. That arg is for custom code handlers only such as timer replacement weak auras or other addons like player which need unique timer guids (that's what timerId is basically). for WA general triggers though. Just need spellid and message like BW timer start

Long term:
1. After BW and DBM have gotten more unified (on same page), merge callbacks together
2. Create an even bigger emphasis to stop using msg text scanning when we're both providing ways to avoid doing msg text scanning. There will probably still be super niche cases it's still needed but typically, they're over used when they shouldn't be and not only does it cause conflicts if boss mods don't match timer/alert text, but also in multi language support.

### PR DESCRIPTION
**Pull request by mrbuds**
# Description

This PR adds new triggers working with both DBM & BigWigs, which fixes #4502

Boss Mod Announce
Boss Mod Stage
Boss Mod Timer

All code for DBM & BW triggers were moved from `Prototypes.lua` & `GenericTrigger.lua` to `BossMods.lua`

New public function `WeakAuras.GetBossStage`

It has a few deprecated functions kept for compatibility with custom code

```
WeakAuras.GetAllDBMTimers
WeakAuras.GetDBMTimerById
WeakAuras.GetDBMTimer
WeakAuras.GetBigWigsTimerById
WeakAuras.GetAllBigWigsTimers
WeakAuras.GetBigWigsStage
WeakAuras.RegisterBigWigsTimer
WeakAuras.RegisterDBMCallback
```

Boss Mod Timer doesn't carry a "type" option, as both bossmod don't have common type options, and will always ignore matches for a cast bar.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested
- [x] /pull 5 (both)
- [x] create test bars (both)
- [x] with jods timeline custom aura (both)
- [x] used in aberrus (BW)
- [ ] used in aberrus (DBM)

### CODE REVIEW

**Comment by MysticalOS** [comment]
don't existing bw triggers have type. or at least a checkbox for cast/not cast? i feel we still have to match this or auraa that rely on it won't move over to unified. i know bw doesn't support advanced typing so it can't really go beyond cd and cast but that's probably fine for now. if it helps i can also just add a simp type arg that just returns true/false in our handler for caatbar or not but looking at this code you basically have that anyways by string finding case "cast" already.

if i'm mistaken this on this i'll correct post when i get home. family keeps forgetting my "stop scheduling shit on tuesday's" rule :/

**Comment by MysticalOS** [comment]
I feel that if there isn't parity with feature of old trigger, no one will use new one at all. The type option is used extensively in many of weak aura examples i gave that targeted BW triggers, cause they need to know whether it's a CD or cast bar and they do trigger off cast bars sometimes.

I can understand no migration since that's extra work, but to ensure people even use the new trigger in future tiers, it has to do everything old one does, or they'd just always use old one. Parity to old BW trigger should be at base a minimum. I can understand why DBMs more thorough advanced typing isn't supported (cause BW obviously doesn't support it). but outs does allow very easily mapping to existing BW cast/nocast trigger mapping.

**Comment by mrbuds** [comment]
It's not a problem of "extra work", we won't do anything that break previously made auras in the middle of an expansion.

**Comment by mrbuds** [comment]
> no i get that, that's why no migration. existing WAs shouldn't be changed now I agree. But i'm saying if 10.2 hits ptr, and someone is creating NEW auras, they're not gonna use the new trigger if it lacks features old one has.

We can ask Ora to check on wago, but i think 99.99% of auras using these triggers doesn't use type options

**Comment by MysticalOS** [comment]
if that's case, we can table the "type" stuff for now until it's further simplified and streamlined.

I just pushed https://github.com/DeadlyBossMods/DBM-Unified/commit/491d8a5ec0e709bae51110c43e81d0ebe62d2198 to address the over complexity of DBMs type system. Whether or not BW does parity to it, at very least, it still makes it easier to parse ours.

File: WeakAuras/BossMods.lua
Comment (Mitalie) [inline WeakAuras/BossMods.lua:397]: DBM_TimerStopAll no longer exists.

File: WeakAuras/BossMods.lua
Comment (Mitalie) [inline WeakAuras/BossMods.lua:1298]: BossMod_TimerStopAll doesn't exist.

File: WeakAuras/BossMods.lua
Comment (Mitalie) [inline WeakAuras/BossMods.lua:721]: Boss Mod Announce trigger expects message and icon as first and second arg of BossMod_Announce (as in DBM_Announce), but BigWigs_Message has them as third and fifth arg, respectively.

File: WeakAuras/BossMods.lua
Comment (Mitalie) [inline WeakAuras/BossMods.lua:1174]: Since this check runs on load, it might be a good idea to add `BigWigs` and `DBM-Core` to OptionalDeps.

File: WeakAuras/BossMods.lua
Comment (Mitalie) [inline WeakAuras/BossMods.lua:1508]: While the no-op funcs prevent errors due to calling nonexistent functions, custom coded aura calling them only works if the events it expects have been registered by other triggers. Ignore if that's intentional.

**Comment by MysticalOS** [comment]
Oh yeah, DBM_TimerStopAll was deprecated long ago when it was ensured TimerStop fires for every timer no matter what. even when mod:stop is called any timer that's stopped fires the correct stop callback per object.

**Comment by Mitalie** [comment]
It doesn't necessarily have to be in this PR, especially if announce triggers are rarely used, but DBM Announce and Boss Mod Announce triggers could be improved by adding support for spellID / key and count, reducing or removing the need to use locale-sensitive messages as trigger options. DBM_Announce has spellID as 4th and count as 7th arg, BigWigs_Message has key as 2nd arg and count would need to be parsed from message similarly to BigWigs_StartBar.

Also, not a new issue but BigWigs Message trigger has BigWigs Addon option to test against the first arg of BigWigs_Message, however the first arg is a table value (<https://github.com/search?q=repo%3ABigWigsMods%2FBigWigs%20BigWigs_Message&type=code>) which will never match the string-valued option. It should probably be removed, or test against `<table>.moduleName` instead.

**Comment by MysticalOS** [comment]
announce triggers are heavily used on echo of neltharian for bomb assignments, specifically for hearts/bombs. But the interesting note of them is they used string matching even when spellid and count was available in the BW announce object. I intentionally changed DBMs strings to match perfectly so they'd work in a unified. but I also didn't realize that weak auras never actually added support for spellid or count in the dbm announce object (and thus not in the boss mod one either). I'd argue it's definitely used enough to warrant addition before merge, or at least a new PR put in immediately after (we can do that if it helps).

**Comment by Mitalie** [comment]
`Boss Mod Timer` trigger passes `extraArg1`/`"noCastBar"` to `Generic:TimerMatches` as the 6th argument, but `DBM:TimerMatches` expects it in the 7th position. The unexpected 6th arg causes `TimerMatches` to always return false.

The above bug would cause `Boss Mod Timer` to never activate when backed by DBM, if not for the call to `Generic:GetTimer` omitting `extraArg1`. While this allows the trigger to function at least partially, the discrepancy doesn't seem intentional.

**Comment by mrbuds** [comment]
@Mitalie thanks for your review! I pushed a commit that should fix the issues you found

**Review by Mitalie — COMMENTED** [review]
Found a few more things.

You can go ahead and mark my first review's comments as resolved - apparently I can't do that myself, it's restricted to the PR submitter and people with write access to the repository.

File: WeakAuras/BossMods.lua
Comment (InfusOnWoW) [inline WeakAuras/BossMods.lua:1395]: I dislike that the generic trigger has to check which of the boss mods is active.

I'd much prefer if the BossMods "modules" would provide a function, e.g. CooldownTimerMatches(...), that behaved in the desired way and had the same api for both dbm and bw. So that the trigger has to not care about any differences.

File: WeakAuras/BossMods.lua
Comment (QartemisT) [reply WeakAuras/BossMods.lua:179]: We can likely expose the colorVars table in DBT.

Though also, you should probably be doing `barOptions['StartColor' .. colorVariables[dbmType] .. 'G']` etc., rather than a nasty switch mess?

Comment (QartemisT) [reply WeakAuras/BossMods.lua:179]: https://github.com/DeadlyBossMods/DBM-Unified/commit/08166b43c2f031de988597ffaab21c5fd74f3b75 Should help with this =]

File: WeakAuras/BossMods.lua
Comment (InfusOnWoW) [reply WeakAuras/BossMods.lua:1598]: On second thought, we probably shouldn't immediately deprecate it. Lots of people use the timeline aura, and won't update soonish.

Though we should still move these forwards to the aura envrionment's FakeWeakAurasMixin.

File: WeakAuras/BossMods.lua
Comment (MysticalOS) [reply WeakAuras/BossMods.lua:179]: I'd like to add that these are NOT option types, just literally color assignments, type is an entirely differet arg in the actual callback, these should never be used to assign type to a weak aura, ever. In fact I just refactored them to do away with them being types in the GUI and now they're just color 1 through color 8 cause users had same confusion. thinking it set a timer type when all it sets is bar color. The types were just what that color was assigned to by default and nothing more. In fact, for this PR, i don't even see how they're relevant at all since what color DBM makes a bar isn't used in these triggers and as far as I've seen type is also omitted. if type is added though then refer to https://github.com/WeakAuras/WeakAuras2/pull/4537#issuecomment-1642607655

File: WeakAuras/BossMods.lua
Comment (InfusOnWoW) [inline WeakAuras/BossMods.lua:1212]: There is a subtle error here. The BigWigs implementation requires a self parameter.

File: WeakAuras/BossMods.lua
Comment (InfusOnWoW) [reply WeakAuras/BossMods.lua:179]: The 'dbmType' is a misnaming on our end, it is what is called colorType in the dbm sources, so yes that's about the color not about the timer type.

**Comment by MysticalOS** [comment]
`local barOptions = DBT.Options or DBM.Bars.options`
Pretty sure we eliminated DBM.Bars last expansion. No one should be running that, the force disable in DBM would make sure DBM for 9.x versions wasn't loading

**Comment by MysticalOS** [comment]
I'd also still love to see Count arg added to "BossMod_Announce" object since it's a single line of parsing it from text (similar to what you already do for BW object) and using the count arg DBM provides (no need to even parse it) Then it's a knock out of the park in full usability next tier

**Comment by mrbuds** [comment]
> I'd also still love to see Count arg added to "BossMod_Announce" object since it's a single line of parsing it from text (similar to what you already do for BW object) and using the count arg DBM provides (no need to even parse it) Then it's a knock out of the park in full usability next tier

Afaik BigWigs_Message event doesn't have a count arg, neither of the 3 message/announce triggers have a count option

**Comment by MysticalOS** [comment]
right, that's why all the weak aura packs in aberus were using string scanning on echo of nelth.

if you look at any of popular packs they were scanning for msg match = "bomb (1)" "bomb (2)" etc. since there was no means to go "if spellid = 12345 and count = 4" on announce.

but yeah of original objects don't even have it then maybe worth another separate PR/issue then and not this one. :)

**Comment by mrbuds** [comment]
i have not seen a single aura for neltharion using announce/message triggers, only bigwigs/dbm timers, but i didn't look at all of them

**Comment by MysticalOS** [comment]
https://wago.io/Y2peTP20U
This one
Screenshot 2023-08-07 at 9 21 36 PM

**Comment by MysticalOS** [comment]
Fortunately it had very few installs. So I'd call it a niche case, but it still exposes the design flaw of that currently this actually is the ONLY way to cherry pick specific counts (in announce object at least, timers support count).

**Review by InfusOnWoW — APPROVED** [review]

**Comment by InfusOnWoW** [comment]
We can certainly add that, though it doesn't need to be part of this PR. So let's merge this first.

**Pull request merged**
